File: lagom-jpa-db2-example/hello-impl/src/main/java/com/lightbend/lagom/hello/impl/HelloServiceImpl.java
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.hello.impl;

import akka.Done;
import akka.NotUsed;
import com.lightbend.lagom.hello.api.GreetingMessage;
import com.lightbend.lagom.hello.api.HelloService;
import com.lightbend.lagom.hello.api.UserGreeting;
import com.lightbend.lagom.hello.impl.entity.HelloCommand;
import com.lightbend.lagom.hello.impl.entity.HelloCommand.Hello;
import com.lightbend.lagom.hello.impl.entity.HelloCommand.UseGreetingMessage;
import com.lightbend.lagom.hello.impl.entity.HelloEntity;
import com.lightbend.lagom.hello.impl.readside.Greetings;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRef;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;
import org.pcollections.PSequence;

import javax.inject.Inject;
import java.util.Optional;

/**
 * Implementation of the HelloService.
 */
public class HelloServiceImpl implements HelloService {
    private final PersistentEntityRegistry persistentEntityRegistry;
    private final Greetings greetings;

    @Inject
    public HelloServiceImpl(PersistentEntityRegistry persistentEntityRegistry, Greetings greetings) {
        this.persistentEntityRegistry = persistentEntityRegistry;
        this.greetings = greetings;
        persistentEntityRegistry.register(HelloEntity.class);
    }

    @Override
    public ServiceCall<NotUsed, String> hello(String id) {
        return request -> {
            // Look up the hello world entity for the given ID.
            PersistentEntityRef<HelloCommand> ref = persistentEntityRegistry.refFor(HelloEntity.class, id);
            // Ask the entity the Hello command.
            return ref.ask(new Hello(id, Optional.empty()));
        };
    }

    @Override
    public ServiceCall<GreetingMessage, Done> useGreeting(String id) {
        return request -> {
            // Look up the hello world entity for the given ID.
            PersistentEntityRef<HelloCommand> ref = persistentEntityRegistry.refFor(HelloEntity.class, id);
            // Tell the entity to use the greeting message specified.
            return ref.ask(new UseGreetingMessage(request.message));
        };

    }

    @Override
    public ServiceCall<NotUsed, PSequence<UserGreeting>> allGreetings() {
        return request -> greetings.all();
    }
}

File: lagom-mq-example-scala/hello-impl/src/main/scala/com/example/hello/impl/HelloServiceImpl.scala
package com.example.hello.impl

import akka.Done
import com.example.hello.api
import com.example.hello.api.HelloService
import com.example.hello.impl.jms.HelloJmsSender
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.api.broker.Topic
import com.lightbend.lagom.scaladsl.broker.TopicProducer
import com.lightbend.lagom.scaladsl.persistence.{EventStreamElement, PersistentEntityRegistry}

import scala.concurrent.ExecutionContext

/**
  * Implementation of the HelloService.
  */
class HelloServiceImpl(
    persistentEntityRegistry: PersistentEntityRegistry,
    jmsSender: HelloJmsSender)(
    implicit ec: ExecutionContext) extends HelloService {

  override def hello(id: String) = ServiceCall { _ =>
    // Look up the Hello entity for the given ID.
    val ref = persistentEntityRegistry.refFor[HelloEntity](id)

    // Ask the entity the Hello command.
    ref.ask(Hello(id))
  }

  override def useGreeting(id: String) = ServiceCall { request =>
    // We've been asked to change the greeting. Send the update information
    // over JMS. (The update will eventually be processed by the
    // HelloJmsReceiverActor.)
    jmsSender.sendGreetingUpdate(id, request.message)
  }
  override def greetingsTopic(): Topic[api.GreetingMessageChanged] =
    TopicProducer.singleStreamWithOffset {
      fromOffset =>
        persistentEntityRegistry.eventStream(HelloEvent.Tag, fromOffset)
          .map(ev => (convertEvent(ev), ev.offset))
    }

  private def convertEvent(helloEvent: EventStreamElement[HelloEvent]): api.GreetingMessageChanged = {
    helloEvent.event match {
      case GreetingMessageChanged(msg) => api.GreetingMessageChanged(helloEvent.entityId, msg)
    }
  }
}
